controller-server/src/main/java/com/yahoo/vespa/hosted/controller/persistence/DeploymentQueue.java

```diff
@@ -0,0 +1,157 @@
+                            .forEach(queue::remove));
+
+        return builder.build();
+    }
+
+    /** Remove all enqueued jobs for the given application. */
+    public void removeJobs(ApplicationId applicationId) {
+        for (JobType jobType : JobType.values())
+            locked(jobType, queue -> {
+                queue.removeIf(triggering -> triggering.applicationId().equals(applicationId));
+            });
+    }
+
+    /** Lock the job queues and read, modify, and store the queue for the given job type. */
+    private void locked(JobType jobType, Consumer<Collection<Triggering>> modifications) {
+        try (Lock lock = curator.lockJobQueues()) {
+            List<Triggering> queue = new CopyOnWriteArrayList<>(curator.readJobQueue(jobType));
+            modifications.accept(queue);
+            curator.writeJobQueue(jobType, queue);
+        }
+    }
+
+    private static boolean isCapacityConstrained(JobType jobType) {
+        return jobType == JobType.stagingTest || jobType == JobType.systemTest;
+    }
+
+    private Optional<BuildJob> toBuildJob(ApplicationId applicationId, JobType jobType) {
+        return controller.applications().get(applicationId)
+                .flatMap(application -> application.deploymentJobs().projectId())
+                .map(projectId -> new BuildJob(projectId, jobType.jobName()));
+    }
+
+
+    static class Triggering implements Comparable<Triggering> {
+
+        private final ApplicationId applicationId;
+        private final Instant at;
+        private final boolean forced;
+        private final boolean applicationVersionUpgrade;
+        private final boolean retry;
+
+        Triggering(ApplicationId applicationId, Instant at, boolean forced, boolean applicationVersionUpgrade, boolean retry) {
+            this.applicationId = requireNonNull(applicationId);
+            this.at = requireNonNull(at);
+            this.forced = forced;
+            this.applicationVersionUpgrade = applicationVersionUpgrade;
+            this.retry = retry;
+        }
+
+        ApplicationId applicationId() { return applicationId; }
+        Instant at() { return at; }
+        boolean forced() { return forced; }
+        boolean applicationVersionUpgrade() { return applicationVersionUpgrade; }
+        boolean retry() { return retry; }
+
+        @Override
+        public int compareTo(Triggering o) {
+            if (forced && ! o.forced) return -1;
+            if (applicationVersionUpgrade && ! o.applicationVersionUpgrade) return -1;
+            if (retry && ! o.retry) return -1;
+            return at.compareTo(o.at);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if ( ! (o instanceof Triggering)) return false;
+
+            Triggering that = (Triggering) o;
+
+            return applicationId.equals(that.applicationId);
```

Comment:
Shouldn't the two build jobs, _app1 Vespa upgrade_ and _app1 application change_ always be considered unequal?

Reply:
No, because this is only a triggering of the external build job, and this triggering carries none of that information. What happens at deployment is determined by the controller. 

However, I already made a note to myself that the queue should choose the entry with the highest priority, whenever an attempt was made to `add` something that was already enqueued.